docs/architecture.md

```diff
@@ -0,0 +1,292 @@
+- `spark.armada.Executor.trackerTimeout` - Specifies the time to
+    wait for the minimum number of Executors.
+- `spark.armada.lookouturl` - Sets the base URL to use for Armada's Lookout
+    service.
+- `spark.armada.health.checkTimeout` - Time to wait for Armada's health check
+    to succeed in seconds.
+- `spark.armada.clusterSelectors` - A comma-separated list of kubernetes
+    label selectors (in key=value format) to ensure the spark Driver and
+    its Executors are deployed to the same cluster.
+- `spark.armada.scheduling.nodeUniformityLabel` - A single kubernetes label to
+    apply to both Driver and Executors.
+- `spark.armada.DriverServiceNamePrefix` - Defines the Driver's service name
+    prefix within Armada. Must consist of lowercase a-z and '-' characters only.
+- `spark.armada.global.labels` - A comma-separated list of kubernetes labels
+    (in key=value format) to be added to all both the Driver and Executor pods.
+- `spark.armada.Driver.labels` - A comma-separated list of kubernetes labels
+    (in key=value format) to be added to the Driver pod.
+- `spark.armada.Executor.labels` A comma-separated list of kubernetes labels
+    (in key=value format) to be added to all Executor pods.
+
+# Building `armada-spark`
+
+## Scala Source
+
+`armada-spark` can be built from source using maven. See the [Building Armada Spark](https://github.com/armadaproject/armada-spark/blob/master/README.md#building-armada-spark)
+section of the README for more information.
+
+## Docker image
+
+`armada-spark` uses docker to produce container images that serve as spark
+Drivers and Executors. See the [Building Docker Images](https://github.com/ClifHouck/armada-spark/blob/master/README.md#building-docker-images)
+section of the README for more information.
+
+# Architecture & Design
+
+## Overview
+
+Fundamentally, `armada-spark`'s architecture and design is closely related to
+the existing Spark Kubernetes cluster manager, but with the added layer of
+Armada and its job system on top.
+
+A great overview of what a "cluster" is from Spark's perspective can be found
+[here](https://spark.apache.org/docs/latest/cluster-overview.html).
+
+```
+// TODO: armada-spark Architecture overview diagram.
+```
+
+The basic flow of an `armada-spark` execution is as follows:
+
+1. End-user calls `ArmadaSparkSubmit.main` via `spark-submit` or `spark-class`.
+2. `ArmadaSparkSubmit`:
+    - Constructs an `ArmadaClient` to communicate with Armada Server via gRPC.
+    - Verifies connectivity and service health of specified Armada URL.
+    - Submits a Driver job to Armada.
+    - Submits `n` Executor jobs to Armada.
+3. After job submissions, `ArmadaSparkSubmit` prints a Lookout link and exits.
+4. Armada schedules Driver and Executor jobs submitted by `ArmadaSparkSubmit`.
+    - Each job submits a pod with a single container.
+    - By using a common label, we can guarantee Driver and Executor jobs are
```

Comment:
Is this still needed now that we are using gang scheduling?

Reply:
I still think it's important to call out. Whether or not we use gang scheduling, we depend on each job to schedule to nodes that share a common `label=value`.